bbq/compiler/peephole_pass.go

```diff
@@ -0,0 +1,170 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright Flow Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package compiler
+
+import (
+	"sort"
+
+	"github.com/onflow/cadence/bbq/opcode"
+	"github.com/onflow/cadence/interpreter"
+)
+
+type BytecodeShiftPair struct {
+	Offset int
+	Shift  int
+}
+
+type PeepholeOptimizer[E, T any] interface {
+	Optimize(code []E) []E
+}
+
+type PeepholeInstructionStaticTypeOptimizer struct {
+	compiler *Compiler[opcode.Instruction, interpreter.StaticType]
+	// bytecodeShifts is a list of where bytecode has been changed
+	// and how much the offset has changed. this is used to patch jumps
+	// after the peephole optimizations have been applied
+	bytecodeShifts []BytecodeShiftPair
+
+	// jumps is a list of offsets in the optimized bytecode that need to be patched
+	jumps []int
+
+	// jumpTargets is a set of offsets in the original bytecode that are jump targets
+	jumpTargets map[int]struct{}
+}
+
+// look up table to improve pattern matching performance
+var PatternsByOpcode map[opcode.Opcode][]PeepholePattern
+
+func init() {
+	PatternsByOpcode = make(map[opcode.Opcode][]PeepholePattern)
+	for _, pattern := range AllPatterns {
+		if PatternsByOpcode[pattern.Opcodes[0]] == nil {
+			PatternsByOpcode[pattern.Opcodes[0]] = make([]PeepholePattern, 0)
+		}
+		PatternsByOpcode[pattern.Opcodes[0]] = append(PatternsByOpcode[pattern.Opcodes[0]], pattern)
```

Comment:
Extract the repeated expression, and the `nil` check / slice allocation is not needed, the `map` lookup returns `nil` by default, and `append` accepts that as the first argument:

```suggestion
		firstOpcode := pattern.Opcodes[0]
		PatternsByOpcode[firstOpcode] = append(PatternsByOpcode[firstOpcode], pattern)
```